miniapps/dpg/util/complexweakform.cpp

```diff
@@ -0,0 +1,1079 @@
+      mat_e_i->owns_blocks = 0;
+      for (int i = 0; i<nblocks; i++)
+      {
+         for (int j = 0; j<nblocks; j++)
+         {
+            SparseMatrix * tmp_r = &mat_e_r->GetBlock(i,j);
+            SparseMatrix * tmp_i = &mat_e_i->GetBlock(i,j);
+            if (Pt->IsZeroBlock(i,i))
+            {
+               PtAe_r->SetBlock(i,j,tmp_r);
+               PtAe_i->SetBlock(i,j,tmp_i);
+            }
+            else
+            {
+               delete tmp_r;
+               delete tmp_i;
+            }
+         }
+      }
+      delete mat_e_r;
+      delete mat_e_i;
+      mat_e_r = PtAe_r;
+      mat_e_i = PtAe_i;
+   }
+   delete Pt;
+
+   mat_r = mfem::Mult(*PtA_r, *P);
+   mat_i = mfem::Mult(*PtA_i, *P);
+
+   PtA_r->owns_blocks = 0;
+   PtA_i->owns_blocks = 0;
+   for (int i = 0; i<nblocks; i++)
+   {
+      for (int j = 0; j<nblocks; j++)
+      {
+         SparseMatrix * tmp_r = &PtA_r->GetBlock(j,i);
+         SparseMatrix * tmp_i = &PtA_i->GetBlock(j,i);
+         if (P->IsZeroBlock(i,i))
+         {
+            mat_r->SetBlock(j,i,tmp_r);
+            mat_i->SetBlock(j,i,tmp_i);
+         }
+         else
+         {
+            delete tmp_r;
+            delete tmp_i;
+         }
+      }
+   }
+   delete PtA_r;
+   delete PtA_i;
+
+   if (mat_e_r)
+   {
+      BlockMatrix *PtAeP_r = mfem::Mult(*mat_e_r, *P);
+      BlockMatrix *PtAeP_i = mfem::Mult(*mat_e_i, *P);
+      mat_e_r->owns_blocks = 0;
+      mat_e_i->owns_blocks = 0;
+      for (int i = 0; i<nblocks; i++)
+      {
+         for (int j = 0; j<nblocks; j++)
+         {
+            SparseMatrix * tmp_r = &mat_e_r->GetBlock(j,i);
+            SparseMatrix * tmp_i = &mat_e_i->GetBlock(j,i);
+            if (P->IsZeroBlock(i,i))
+            {
+               PtAeP_r->SetBlock(j,i,tmp_r);
+               PtAeP_i->SetBlock(j,i,tmp_i);
+            }
+            else
+            {
+               delete tmp_r;
+               delete tmp_i;
+            }
+         }
+      }
+
+      delete mat_e_r;
+      delete mat_e_i;
+      mat_e_r = PtAeP_r;
+      mat_e_i = PtAeP_i;
+   }
+   height = 2*mat_r->Height();
+   width = 2*mat_r->Width();
+}
+
+/// Assembles the form i.e. sums over all domain integrators.
+void ComplexDPGWeakForm::Assemble(int skip_zeros)
+{
+   ElementTransformation *eltrans;
+   Array<int> faces, ori;
+
+   DofTransformation * doftrans_i, *doftrans_j;
+   if (mat_r == NULL)
+   {
+      AllocMat();
+   }
+
+   // loop through the elements
+   int dim = mesh->Dimension();
+   DenseMatrix B_r, Be_r, G_r, Ge_r, A_r;
+   DenseMatrix B_i, Be_i, G_i, Ge_i, A_i;
+   Vector vec_e_r, vec_r, b_r;
+   Vector vec_e_i, vec_i, b_i;
+   Array<int> vdofs;
+
+   // loop through elements
+   for (int iel = 0; iel < mesh -> GetNE(); iel++)
+   {
+      if (dim == 1)
+      {
+         mesh->GetElementVertices(iel, faces);
+      }
+      if (dim == 2)
+      {
+         mesh->GetElementEdges(iel, faces, ori);
+      }
+      else //dim = 3
```

Comment:
This isn't safe against potential higher dimensions.  I know there has been some talk of that going around so it might be better to test for dim==3 and else into an error.

Reply:
done!